src/packagers/JSConcatPackager.js

```diff
@@ -0,0 +1,249 @@
+    this.buffer = '';
+    this.exports = new Map();
+    this.wildcards = new Map();
+    this.moduleMap = new Map();
+    this.needsPrelude = false;
+
+    for (let asset of this.bundle.assets) {
+      // If this module is referenced by another bundle, it needs to be exposed externally.
+      let isExposed = !Array.from(asset.parentDeps).every(dep =>
+        this.bundle.assets.has(this.bundler.loadedAssets.get(dep.parent))
+      );
+
+      if (
+        isExposed ||
+        (this.bundle.entryAsset === asset && this.bundle.parentBundle)
+      ) {
+        this.exposedModules.add(asset);
+        this.needsPrelude = true;
+      }
+
+      for (let mod of asset.depAssets.values()) {
+        if (!this.bundle.assets.has(mod)) {
+          this.needsPrelude = true;
+          break;
+        }
+      }
+    }
+
+    if (this.needsPrelude) {
+      if (this.bundle.entryAsset) {
+        this.exposedModules.add(this.bundle.entryAsset);
+      }
+
+      await this.write(prelude + '(function (require) {\n');
+    } else {
+      await this.write(`
+(function () {
+  function $parcel$exportWildcard(dest, source) {
+    Object.keys(source).forEach(function(key) {
+      if(key === "default" || key === "__esModule") return;
+      Object.defineProperty(dest, key, {
+        enumerable: true,
+        get: function get() {
+          return source[key];
+        }
+      });
+    });
+
+    return dest;
+  }
+`);
+    }
+  }
+
+  getExportIdentifier(asset) {
+    return '$' + asset.id + '$exports';
+  }
+
+  async addAsset(asset) {
+    if (this.addedAssets.has(asset)) {
+      return;
+    }
+
+    this.addedAssets.add(asset);
+    let js = asset.generated.js;
+
+    this.moduleMap.set(asset.id, asset);
+    this.wildcards.set(asset.id, asset.cacheData.wildcards);
+
+    for (let key in asset.cacheData.exports) {
+      let local = '$' + asset.id + '$export$' + asset.cacheData.exports[key];
+      if (key !== local) {
+        this.exports.set(key, local);
+      }
+    }
+
+    for (let [dep, mod] of asset.depAssets) {
+      let depName = '$' + asset.id + '$require$' + t.toIdentifier(dep.name);
+      let moduleName = this.getExportIdentifier(mod);
+
+      // If this module is not in the current bundle, generate a `require` call for it.
+      if (!this.bundle.assets.has(mod)) {
+        moduleName = `require(${mod.id})`;
+      }
+
+      js = js.split(depName).join(moduleName);
+
+      // If this was an ES6 export all (e.g. export * from 'foo'), resolve to the original exports.
+      if (dep.isExportAll) {
+        for (let exp in mod.cacheData.exports) {
+          let id = mod.cacheData.exports[exp];
+          if (id !== 'default') {
+            let key = '$' + asset.id + '$export$' + id;
+            asset.cacheData.exports[key] = id;
+            this.exports.set(
+              key,
+              this.exports.get('$' + mod.id + '$export$' + id) || exp
+            );
+          }
+        }
+      }
+
+      if (dep.isES6Import) {
```

Comment:
let's see if we can move this mess into the babel plugin instead of doing this string manipulation